infobind/cache.go

```diff
@@ -0,0 +1,256 @@
+type BindCache struct {
+	Cache map[string][]*BindData
+}
+
+//Handle hold a atomic bindCache
+type Handle struct {
+	bind atomic.Value
+}
+
+//HandleUpdater use to update the BindCache
+type HandleUpdater struct {
+	Parser         *parser.Parser
+	LastUpdateTime types.Time
+	Ctx            sessionctx.Context
+	*Handle
+}
+
+type bindRecord struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	Status      int64
+	CreateTime  types.Time
+	UpdateTime  types.Time
+	Charset     string
+	Collation   string
+}
+
+//NewHandle create a Handle with a BindCache
+func NewHandle() *Handle {
+	handle := &Handle{}
+	bc := &BindCache{
+		Cache: make(map[string][]*BindData, defaultBindCacheSize),
+	}
+	handle.bind.Store(bc)
+	return handle
+}
+
+//Get get bindCache from a Handle
+func (h *Handle) Get() *BindCache {
+	bc := h.bind.Load()
+	if bc != nil {
+		return bc.(*BindCache)
+	}
+	return &BindCache{
+		Cache: make(map[string][]*BindData, defaultBindCacheSize),
+	}
+}
+
+//Load Diff use to load new bind info to bindCache bc
+func (h *HandleUpdater) loadDiff(sql string, bc *BindCache) error {
+	tmp, err := h.Ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := tmp[0]
+	defer terror.Call(rs.Close)
+
+	fs := rs.Fields()
+	chk := rs.NewChunk()
+	for {
+		err = rs.Next(context.TODO(), chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if chk.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record, err := decodeBindTableRow(row, fs)
+			if err != nil {
+				log.Errorf("row decode error %s", err)
+				continue
+			}
+			log.Infof("record %v", record)
+			err = bc.appendNode(h.Ctx, record, h.Parser)
+			if err != nil {
+				continue
+			}
+
+			if record.UpdateTime.Compare(h.LastUpdateTime) == 1 {
+				h.LastUpdateTime = record.UpdateTime
+			}
+		}
+		chk = chunk.Renew(chk, h.Ctx.GetSessionVars().MaxChunkSize)
+	}
+}
+
+//Update update the HandleUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false
+func (h *HandleUpdater) Update(fullLoad bool) error {
+	var (
+		err error
+		sql string
+	)
+	bc := h.Get()
+	if fullLoad {
+		sql = fmt.Sprintf("select * from mysql.bind_info")
+	} else {
+		sql = fmt.Sprintf("select * from mysql.bind_info where update_time > \"%s\"", h.LastUpdateTime.String())
+	}
+	err = h.loadDiff(sql, bc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h.bind.Store(bc)
+	bc.Display()
+	return nil
+}
+
+func parseSQL(sctx sessionctx.Context, parser *parser.Parser, sql string, charset string, collation string) ([]ast.StmtNode, []error, error) {
+	return parser.Parse(sql, charset, collation)
+}
+
+func decodeBindTableRow(row chunk.Row, fs []*ast.ResultField) (bindRecord, error) {
+	var value bindRecord
+	for i, f := range fs {
+		switch {
+		case f.ColumnAsName.L == "original_sql":
```

Comment:
Why not use column offset?

Reply:
What the diffence between this columnAsName with column offset?

Reply:
You do not need to check the name, just use write `a = row.get(0), b = row.get(1)`, it is simpler.

Reply:
ok